### ISSUE TITLE
Database is locked

### ISSUE THREAD

**Issue by etwatson** [juliema/aTRAM#267]
Hi, everything was working last week, but today I'm getting this error. I deleted the old database and re-ran atram_preprocessor.py to create a new one, but that still produced this error. 
```
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-26 11:02:18 INFO: ################################################################################
2019-09-26 11:02:18 INFO: aTRAM version: v2.3.0
2019-09-26 11:02:18 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-09-26 11:02:18 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-26 11:02:18 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06058_1.fasta", iteration 1
2019-09-26 11:02:18 INFO: Blasting query against shards: iteration 1
2019-09-26 11:02:30 ERROR: Exception: database is locked
2019-09-26 11:02:31 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06355_2.fasta", iteration 1
2019-09-26 11:02:31 INFO: Blasting query against shards: iteration 1
^C

**Comment by rafelafrance** [comment]
I see you are running v2.3.0. The last official release was v2.2.0. Would you try checking out that version and rerunning:
```
git checkout v2.2.0
```

**Comment by etwatson** [comment]
I get the same error with v2.2.0:
```
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 10:40:50 INFO: ################################################################################
2019-09-30 10:40:50 INFO: aTRAM version: v2.2.0
2019-09-30 10:40:50 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-09-30 10:40:50 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 10:40:50 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06058_1.fasta", iteration 1
2019-09-30 10:40:50 INFO: Blasting query against shards: iteration 1
2019-09-30 10:41:05 ERROR: Exception: database is locked
2019-09-30 10:41:06 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06355_2.fasta", iteration 1
2019-09-30 10:41:06 INFO: Blasting query against shards: iteration 1
^C

**Comment by etwatson** [comment]
I checked out v2.2.0 and reran the atram_preprocessor.py script on my read files. However, I still get the error when running atram.py:
```
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 11:34:14 INFO: ################################################################################
2019-09-30 11:34:14 INFO: aTRAM version: v2.2.0
2019-09-30 11:34:14 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-09-30 11:34:14 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 30 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 11:34:14 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06058_1.fasta", iteration 1
2019-09-30 11:34:14 INFO: Blasting query against shards: iteration 1
2019-09-30 11:34:30 ERROR: Exception: database is locked
2019-09-30 11:34:32 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06355_2.fasta", iteration 1
2019-09-30 11:34:32 INFO: Blasting query against shards: iteration 1
^C

**Comment by etwatson** [comment]
It only works if I reduce --cpus to 1. Am I doing something dumb here?
``` python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 1 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 12:06:58 INFO: ################################################################################
2019-09-30 12:06:58 INFO: aTRAM version: v2.2.0
2019-09-30 12:06:58 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-09-30 12:06:58 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o ovary_limited -Q /media/mcclintock/etw/gonadRNA/SLG_assembly/SD.ovary_limited.fa --assembler spades --cpus 1 --max-memory 100 --ins-length 400 --iterations 20 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-09-30 12:06:58 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.ovary_limited_SD_TCALIF_06058_1.fasta", iteration 1
2019-09-30 12:06:58 INFO: Blasting query against shards: iteration 1
2019-09-30 12:08:05 INFO: All 74 blast results completed
2019-09-30 12:08:05 INFO: 642 blast hits in iteration 1
2019-09-30 12:08:05 INFO: Writing assembler input files: iteration 1
2019-09-30 12:08:05 INFO: Assembling shards with spades: iteration 1
2019-09-30 12:08:11 INFO: Saving assembled contigs: iteration 1
2019-09-30 12:08:11 INFO: Creating new query files: iteration 1
2019-09-30 12:08:11 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.ovary_limited_SD_TCALIF_06058_1.fasta_01_long_reads.fasta", iteration 2
2019-09-30 12:08:11 INFO: Blasting query against shards: iteration 2
^C

**Comment by rafelafrance** [comment]
This is interesting. So, I'm guessing that you have 12 cores on your machine. This may have given me something to test. I will see if I can replicate this issue with:
- Adding more cores (--cpus) than the machine has
- And an older version of python

Thanks

**Comment by etwatson** [comment]
My machine has two 10 core Xeon processors, but I can not run --cpus >12 without getting the "Database is locked" error.

**Comment by etwatson** [comment]
FYI, those blastn errors are due to malformed fasta in the query.

**Comment by rafelafrance** [comment]
I have not been able to replicate the issue locally, however, I do have a patch for you to try. Please do
```
git checkout master
git pull
```

**Comment by etwatson** [comment]
Thank you for looking into this. The program is amazing, and I am definitely getting some  I think the issue may have originated from running atram.py in a bash loop and then killing the processes with ^c. Something with sqlite3 doesn't like it. I can reproduce the sqlite3 error in this way, then all subsequent atram.py jobs on that shell crash with the "Database is locked" error. Just switching to a new shell screen fixes the error and I am able to run with --cpus 10.

**Comment by rafelafrance** [comment]
I'll try & replicate the issue this Thursday. I only work on aTRAM one day a week

**Comment by rafelafrance** [comment]
I still cannot replicate the issue. I found an older machine with Python 3.5.2 on it and interrupted a batch run via ctrl-c several times during the run. The database didn't lock. I need to think about what the race condition may be. I'm open to ideas.

Some things to look at:

- Who has the DB locked with `fuser path/to/sqlite/database`. I'd expect this to be python but you should confirm this.
- Did the DB become corrupted? `sqlite3 path/to/my.sqlite.db "pragma integrity_check;"`  Look at both the main DB you created via the preprocessor and the temporary DB located in the saved temp dir `temp/db/*_temp.sqlite.db`
- Which reminds me that another source of problems, if you have sqlite3 installed via apt and it's incompatible with the python version. A real long shot.

Just in case my tests do not replicate your scripts here's what I ran.

Here's my bash script
```bash
for i in  1 2 3 4 5 6 7 8
do
  echo "Run $i"
  python3 atram.py @args/db_locked.args --output-prefix=output/ptgor.$i --log-file=output/ptgor.$i.log
done
```

And here are the arguments in db_locked.args
```
--query=query/Phum.PHUM003340-PA.pep.fasta
--blast-db=db/ptgor
--assembler=spades
--temp-dir=temp
--keep-temp-dir
--cpus=20
```

**Comment by etwatson** [comment]
Thanks for trying. I downloaded the recent patch, but now the util_check_requirements.py shows that some things aren't installed even though they are. I don't have abyss installed, but numpy, psutil, python-dateutil and pytz are all installed despite the output.

```
python3.6 util_check_requirements.py 
Python version.......................... OK
makeblastdb............................. OK
tblastn................................. OK
blastn.................................. OK
abyss................................... FAIL
trinity................................. OK
velvet.................................. OK
spades.................................. OK
bwa..................................... OK
bowtie2................................. OK
exonerate............................... OK
biopython............................... OK
numpy................................... FAIL
psutil.................................. FAIL
python-dateutil......................... FAIL
pytz.................................... FAIL
six..................................... OK

Results:
  atram.py and atram_preprocessor.py will not work because the psutil module is missing (install it with pip)
  atram.py and atram_preprocessor.py will not work because the numpy module is missing (install it with pip)
  atram.py will work but you are missing the abyss assembler
```

**Comment by rafelafrance** [comment]
Fix requirements should be checking versions >=, but I'll make sure. Just checking, did you run this utility from the virtual environment?

The stitcher bug is something that should be easy to fix.

I'll work on both Friday when I'm back on aTRAM.

**Comment by rafelafrance** [comment]
Stitcher should be fixed.

I've updated the version parsing to use a more standard function.

**Comment by etwatson** [comment]
The problem was too big for me to continue on the same machine. The error caused hundreds of dropout genes in various populations, so I applied for an HPC account with my university and was able to install aTRAM on the HPC cluster.

However, I now get the "database locked" on the HPC cluster every single time. I'm at a loss, since the HPC install should have been a clean one. 

```
2019-10-22 07:34:36 INFO: ################################################################################
2019-10-22 07:34:36 INFO: aTRAM version: v2.3.0
2019-10-22 07:34:36 INFO: Python version: 3.7.4 (default, Jul 16 2019, 16:40:45) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
2019-10-22 07:34:36 INFO: /home/rcf-proj2/ew2/ericwats/repo/aTRAM/atram.py --db /home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM --probe /home/rcf-proj2/ew2/ericwats/probes/SD.testes_limited.fa --output-prefix testis --assembler spades --iterations 25 --cpus 40 --temp-dir AB --keep-temp-dir --ins-length 400
2019-10-22 07:34:36 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "SD.testes_limited.fa", iteration 1
2019-10-22 07:34:36 INFO: Blasting query against shards: iteration 1
2019-10-22 07:36:44 ERROR: Exception: database is locked
2019-10-22 07:36:47 INFO: ################################################################################
```

**Comment by etwatson** [comment]
When I run the util_check_requirements.py, I get the following error no the HPC cluster:
```
[ericwats@hpc-login2 aTRAM]$ python3.7 util_check_requirements.py 
Python version.......................... OK
makeblastdb............................. OK
tblastn................................. OK
blastn.................................. OK
abyss................................... FAIL
trinity................................. FAIL
velvet.................................. FAIL
spades.................................. OK
bwa..................................... FAIL
bowtie2................................. FAIL
exonerate............................... OK

Results:
Traceback (most recent call last):
  File "util_check_requirements.py", line 122, in <module>
    report_results()
  File "util_check_requirements.py", line 97, in report_results
    requires('biopython', 'the biopython module is missing (install with pip)')
  File "util_check_requirements.py", line 75, in requires
    if not RESULTS[module]:
KeyError: 'biopython'
```
I thought maybe the biopython install failed, but it doesn't appear so:
```
Requirement already satisfied: biopython in /auto/rcf-40/ericwats/.local/lib/python3.7/site-packages (1.74)
Requirement already satisfied: numpy in /auto/rcf-40/ericwats/.local/lib/python3.7/site-packages (from biopython) (1.17.3)
Requirement already satisfied: psutil in /auto/rcf-40/ericwats/.local/lib/python3.7/site-packages (5.6.3)
```

**Comment by juliema** [comment]
thanks for the info. we will look into this asap.  Rafe I wonder if we want to try this backup database option?
https://www.arysontechnologies.com/blog/fix-sqlite-error-database-locked/

**Comment by etwatson** [comment]
I appreciate the attention you are putting into this. I'm not sure what happened. I downloaded an earlier version of aTRAM2 and it worked perfectly for a while when I was generating preliminary data for this project. I kept aTRAM updated, and when I started the project full scale, it immediately stopped working for me. Unfortunately, I don't recall which version actually worked.

**Comment by rafelafrance** [comment]
AFAICT your environment is sane & nothing in the data jumps out at me as wrong. But I can't replicate the issue either so I'm shooting at the Heisenbug in the dark.

The main DB for atram is built during the preprocessor step. It is read-only after that. Atram itself creates a fresh temp DB for every run. I believe that that is what's getting locked. So a DB backup isn't going to help here.

The main atram thread in atram.py does keep a DB connection open. And the child processes open their own DB connections and pretty much keep them open for as long as their working. I do commits etc. so in theory all should be OK... but clearly it's not.

I'm going to try to manipulate the connections to only have them open for a single query and bump up the time out setting etc. The problem here is that if you have a bazillion blast hits it may still lock up. The chance will be smaller but not non-zero. Then again, none of this should be needed. Maybe I missed a commit somewhere and it's hitting you harder than other users.

PS: The environment checker needs some tweaking too but that's a lower priority for now.

**Comment by etwatson** [comment]
Thanks for looking into it. I don't know why it is hitting me harder than other users, but the all of the users at my university that use the High Performance Computing resources would share the problem.

**Comment by rafelafrance** [comment]
I've bumped up the SQLite3 connection timeout from 5 to 30 seconds and rearranged context  management for the database.

Please give it a try.

I am still using WAL type transactions and I cannot use autocommit mode because it kills performance by orders of magnitude.

**Comment by etwatson** [comment]
Hi, I used ``` git pull``` and tried again on my local machine. On my university's HPC cluster, I also installed the recent version and submitted a job and will edit this comment when it runs. 

On my local machine, I didn't get the "Database is locked" error, so that's positive! I did get an "Exception: file is not a database" error though, and the assemblies failed. 

```
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-25 12:04:07 INFO: ################################################################################
2019-10-25 12:04:07 INFO: aTRAM version: v2.3.0
2019-10-25 12:04:07 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-25 12:04:07 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-25 12:04:07 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-25 12:04:07 INFO: Blasting query against shards: iteration 1
2019-10-25 12:04:40 INFO: All 74 blast results completed
2019-10-25 12:04:40 INFO: 427 blast hits in iteration 1
2019-10-25 12:04:40 INFO: Writing assembler input files: iteration 1
2019-10-25 12:04:42 INFO: Assembling shards with spades: iteration 1
2019-10-25 12:04:43 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__jzdi8kp/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_sz_ruycn/spades' returned non-zero exit status 1.
2019-10-25 12:04:44 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-25 12:04:44 INFO: Blasting query against shards: iteration 1
2019-10-25 12:04:59 ERROR: Exception: file is not a database
2019-10-25 12:05:00 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01780_3.fasta", iteration 1
2019-10-25 12:05:00 INFO: Blasting query against shards: iteration 1
2019-10-25 12:05:28 INFO: All 74 blast results completed
2019-10-25 12:05:28 INFO: 1052 blast hits in iteration 1
2019-10-25 12:05:28 INFO: Writing assembler input files: iteration 1
2019-10-25 12:05:32 INFO: Assembling shards with spades: iteration 1
2019-10-25 12:05:32 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__jzdi8kp/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01780_3.fasta_01_862ijvat/spades' returned non-zero exit status 1.
```
This time, in my ```/tmp/``` directory, I can see that all 74 blast results completed, but no reads were extracted for the spades assembly. The ```spades/``` directory is empty. 

```
-rw-r--r--  1 etw edmands-lab 12072 Oct 25 12:04 AB.aTRAM.074.blast.results.json
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 paired_1.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 paired_2.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_1.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_2.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_any.fasta
drwxr-xr-x  2 etw edmands-lab  4096 Oct 25 12:04 spades/

```

**Comment by juliema** [comment]
ok this is positive! Did you check that the path to the DB and the DB name are correct?

On Fri, Oct 25, 2019 at 12:17 PM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi, I used git pull and tried again on my local machine. On my university's HPC cluster, I also installed the recent version and submitted a job and will edit this comment when it runs.

On my local machine, I didn't get the "Database is locked" error, so that's positive! I did get an "Exception: file is not a database" error though, and the assemblies failed.

python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-25 12:04:07 INFO: ################################################################################
2019-10-25 12:04:07 INFO: aTRAM version: v2.3.0
2019-10-25 12:04:07 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-25 12:04:07 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-25 12:04:07 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-25 12:04:07 INFO: Blasting query against shards: iteration 1
2019-10-25 12:04:40 INFO: All 74 blast results completed
2019-10-25 12:04:40 INFO: 427 blast hits in iteration 1
2019-10-25 12:04:40 INFO: Writing assembler input files: iteration 1
2019-10-25 12:04:42 INFO: Assembling shards with spades: iteration 1
2019-10-25 12:04:43 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__jzdi8kp/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_sz_ruycn/spades' returned non-zero exit status 1.
2019-10-25 12:04:44 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-25 12:04:44 INFO: Blasting query against shards: iteration 1
2019-10-25 12:04:59 ERROR: Exception: file is not a database
2019-10-25 12:05:00 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01780_3.fasta", iteration 1
2019-10-25 12:05:00 INFO: Blasting query against shards: iteration 1
2019-10-25 12:05:28 INFO: All 74 blast results completed
2019-10-25 12:05:28 INFO: 1052 blast hits in iteration 1
2019-10-25 12:05:28 INFO: Writing assembler input files: iteration 1
2019-10-25 12:05:32 INFO: Assembling shards with spades: iteration 1
2019-10-25 12:05:32 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__jzdi8kp/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01780_3.fasta_01_862ijvat/spades' returned non-zero exit status 1.


This time, in my /tmp/ directory, I can see that all 74 blast results completed, but no reads were extracted for the spades assembly. The spades/ directory is empty.

-rw-r--r--  1 etw edmands-lab 12072 Oct 25 12:04 AB.aTRAM.074.blast.results.json
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 paired_1.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 paired_2.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_1.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_2.fasta
-rw-r--r--  1 etw edmands-lab     0 Oct 25 12:04 single_any.fasta
drwxr-xr-x  2 etw edmands-lab  4096 Oct 25 12:04 spades/

**Comment by etwatson** [comment]
Yes, of course the DB is named correctly and present in the indicated folder. I still randomly get the "Database is locked" error, and have not seen the " Exception: file is not a database" error again. 

Unfortunately, I have to move on from aTRAM. It isn't ideal for my project (aTRAM would be the best approach), but it has been a month of stalling and I need to publish. For now I need to turn my focus towards an alternative. I will continue to respond to this issue and help in any way that I can since it is clear that this issue affects many potential users. Thank you for attention to the matter.

```
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: ################################################################################
2019-10-28 10:54:32 INFO: aTRAM version: v2.3.0
2019-10-28 10:54:32 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-28 10:54:32 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 10:54:32 INFO: Blasting query against shards: iteration 1
2019-10-28 10:55:32 ERROR: Exception: database is locked
2019-10-28 10:55:34 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 10:55:34 INFO: Blasting query against shards: iteration 1
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: ################################################################################Exception: file is not a database
2019-10-28 10:54:32 INFO: aTRAM version: v2.3.0
2019-10-28 10:54:32 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-28 10:54:32 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 10:54:32 INFO: Blasting query against shards: iteration 1
2019-10-28 10:55:32 ERROR: Exception: database is locked
2019-10-28 10:55:34 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 10:55:34 INFO: Blasting query against shards: iteration 1
2019-10-28 11:00:27 INFO: ################################################################################
2019-10-28 11:00:27 INFO: aTRAM version: v2.3.0
2019-10-28 11:00:27 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]Exception: file is not a database
2019-10-28 11:00:27 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 11:00:27 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 11:00:27 INFO: Blasting query against shards: iteration 1
2019-10-28 11:00:43 INFO: All 74 blast results completed
2019-10-28 11:00:43 INFO: 427 blast hits in iteration 1
2019-10-28 11:00:43 INFO: Writing assembler input files: iteration 1
2019-10-28 11:00:43 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:00:44 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_5ti_zfqo/spades' returned non-zero exit status 1.
2019-10-28 11:00:45 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 11:00:45 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:01 INFO: All 74 blast results completed
2019-10-28 11:01:01 INFO: 248 blast hits in iteration 1
2019-10-28 11:01:01 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:01 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:01 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01590_2.fasta_01_ynn_x8m5/spades' returned non-zero exit status 1.
2019-10-28 11:01:02 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01780_3.fasta", iteration 1
2019-10-28 11:01:02 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:23 INFO: All 74 blast results completed
2019-10-28 11:01:23 INFO: 1052 blast hits in iteration 1
2019-10-28 11:01:23 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:23 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:23 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01780_3.fasta_01_a23ug0qw/spades' returned non-zero exit status 1.
2019-10-28 11:01:24 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01824_4.fasta", iteration 1
2019-10-28 11:01:24 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:43 INFO: All 74 blast results completed
2019-10-28 11:01:43 INFO: 533 blast hits in iteration 1
2019-10-28 11:01:43 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:43 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:43 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01824_4.fasta_01_r006a35_/spades' returned non-zero exit status 1.
2019-10-28 11:01:44 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01945_5.fasta", iteration 1
2019-10-28 11:01:44 INFO: Blasting query against shards: iteration 1
2019-10-28 11:02:07 INFO: All 74 blast results completed
2019-10-28 11:02:07 INFO: 801 blast hits in iteration 1
2019-10-28 11:02:07 INFO: Writing assembler input files: iteration 1
2019-10-28 11:02:10 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:02:10 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8Exception: file is not a database/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01945_5.fasta_01_fgpr8rgp/spades' returned non-zero exit status 1.
2019-10-28 11:02:11 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01967_6.fasta", iteration 1
2019-10-28 11:02:11 INFO: Blasting query against shards: iteration 1
```

**Comment by juliema** [comment]
so just looking at this output. It looks like spades is not performing optimally. I might try another assembler - trinity or abyss to see if it is an issue with how spades is installed. With 801 blast hits retrieved I would expect an assembly.



On Mon, Oct 28, 2019 at 11:12 AM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

Yes, of course the DB is named correctly and present in the indicated folder. I still randomly get the "Database is locked" error, and have not seen the " Exception: file is not a database" error again.

Unfortunately, I have to move on from aTRAM. It isn't ideal for my project (aTRAM would be the best approach), but it has been a month of stalling and I need to publish. For now I need to turn my focus towards an alternative. I will continue to respond to this issue and help in any way that I can since it is clear that this issue affects many potential users. Thank you for attention to the matter.

python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: ################################################################################
2019-10-28 10:54:32 INFO: aTRAM version: v2.3.0
2019-10-28 10:54:32 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-28 10:54:32 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 10:54:32 INFO: Blasting query against shards: iteration 1
2019-10-28 10:55:32 ERROR: Exception: database is locked
2019-10-28 10:55:34 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 10:55:34 INFO: Blasting query against shards: iteration 1
python3.6 /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: ################################################################################Exception: file is not a database
2019-10-28 10:54:32 INFO: aTRAM version: v2.3.0
2019-10-28 10:54:32 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]
2019-10-28 10:54:32 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 10:54:32 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 10:54:32 INFO: Blasting query against shards: iteration 1
2019-10-28 10:55:32 ERROR: Exception: database is locked
2019-10-28 10:55:34 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 10:55:34 INFO: Blasting query against shards: iteration 1
2019-10-28 11:00:27 INFO: ################################################################################
2019-10-28 11:00:27 INFO: aTRAM version: v2.3.0
2019-10-28 11:00:27 INFO: Python version: 3.6.6 |Anaconda, Inc.| (default, Oct 9 2018, 12:34:16) [GCC 7.3.0]Exception: file is not a database
2019-10-28 11:00:27 INFO: /usr/local/repo/aTRAM/atram.py -b /media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM.sqlite.db -o AB -Q SD.male_carcass_limited.fa --assembler spades --cpus 20 --max-memory 100 --ins-length 400 --iterations 25 --log-file AB.log --temp-dir /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp --keep-temp-dir
2019-10-28 11:00:27 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_00592_1.fasta", iteration 1
2019-10-28 11:00:27 INFO: Blasting query against shards: iteration 1
2019-10-28 11:00:43 INFO: All 74 blast results completed
2019-10-28 11:00:43 INFO: 427 blast hits in iteration 1
2019-10-28 11:00:43 INFO: Writing assembler input files: iteration 1
2019-10-28 11:00:43 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:00:44 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_5ti_zfqo/spades' returned non-zero exit status 1.
2019-10-28 11:00:45 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01590_2.fasta", iteration 1
2019-10-28 11:00:45 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:01 INFO: All 74 blast results completed
2019-10-28 11:01:01 INFO: 248 blast hits in iteration 1
2019-10-28 11:01:01 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:01 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:01 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01590_2.fasta_01_ynn_x8m5/spades' returned non-zero exit status 1.
2019-10-28 11:01:02 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01780_3.fasta", iteration 1
2019-10-28 11:01:02 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:23 INFO: All 74 blast results completed
2019-10-28 11:01:23 INFO: 1052 blast hits in iteration 1
2019-10-28 11:01:23 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:23 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:23 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01780_3.fasta_01_a23ug0qw/spades' returned non-zero exit status 1.
2019-10-28 11:01:24 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01824_4.fasta", iteration 1
2019-10-28 11:01:24 INFO: Blasting query against shards: iteration 1
2019-10-28 11:01:43 INFO: All 74 blast results completed
2019-10-28 11:01:43 INFO: 533 blast hits in iteration 1
2019-10-28 11:01:43 INFO: Writing assembler input files: iteration 1
2019-10-28 11:01:43 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:01:43 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01824_4.fasta_01_r006a35_/spades' returned non-zero exit status 1.
2019-10-28 11:01:44 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01945_5.fasta", iteration 1
2019-10-28 11:01:44 INFO: Blasting query against shards: iteration 1
2019-10-28 11:02:07 INFO: All 74 blast results completed
2019-10-28 11:02:07 INFO: 801 blast hits in iteration 1
2019-10-28 11:02:07 INFO: Writing assembler input files: iteration 1
2019-10-28 11:02:10 INFO: Assembling shards with spades: iteration 1
2019-10-28 11:02:10 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 20 --memory 100 --cov-cutoff off -o /media/mcclintock/etw/gonadRNA/SLG_assembly/tmp/atram__urvc4b8Exception: file is not a database/AB.aTRAM_SD.male_carcass_limited_SD_TCALIF_01945_5.fasta_01_fgpr8rgp/spades' returned non-zero exit status 1.
2019-10-28 11:02:11 INFO: aTRAM blast DB = "/media/mcclintock/seq_data_archive/aTRAM/AB.aTRAM", query = "SD.male_carcass_limited_SD_TCALIF_01967_6.fasta", iteration 1
2019-10-28 11:02:11 INFO: Blasting query against shards: iteration 1

**Comment by juliema** [comment]
when referring to previous issue you just mean returning exit status 1?

will you try abyss?

Will you check that trinity is loaded with the correct version?



On Mon, Oct 28, 2019 at 11:36 AM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

Same issue as before with Spades - Trinity also is failing, but there are no reads for the assembly in the tmp folder:

-rw-r--r-- 1 etw edmands-lab  9569 Oct 28 11:34 BB.aTRAM.091.blast.results.json
-rw-r--r-- 1 etw edmands-lab     0 Oct 28 11:34 paired_1.fasta
-rw-r--r-- 1 etw edmands-lab     0 Oct 28 11:34 paired_2.fasta
-rw-r--r-- 1 etw edmands-lab     0 Oct 28 11:34 single_1.fasta
-rw-r--r-- 1 etw edmands-lab     0 Oct 28 11:34 single_2.fasta
-rw-r--r-- 1 etw edmands-lab     0 Oct 28 11:34 single_any.fasta

**Comment by etwatson** [comment]
> are there reads for the assembly in the tmp folder when running spades?

Correct. There are no reads extracted from the blast results to run spades or trinity -> https://github.com/juliema/aTRAM/issues/267#issuecomment-546478944

The Trinity version I am using is v2.5.1
The Spades version is v3.13.0 
Abyss is the only assembler not installed on this machine.

**Comment by juliema** [comment]
ok just totally checking here did you module load trinity and spades?

the version of trinity should work fine, spades is a different version than what we have tested for. 3.11.1 . -- so there it could be a version issue.

One test. will you remove the --temp-dir and --keep-temp-dir options and try it?



On Mon, Oct 28, 2019 at 12:02 PM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

are there reads for the assembly in the tmp folder when running spades?

Correct. There are no reads extracted from the blast results to run spades or trinity -> #267 (comment)<https://github.com/juliema/aTRAM/issues/267#issuecomment-546478944>

The Trinity version I am using is v2.5.1
The Spades version is v3.13.0
Abyss is the only assembler not installed on this machine.

**Comment by etwatson** [comment]
> ok just totally checking here did you module load trinity and spades?

no, I don't know what that means. 


Still get the assembler error when removing --tmp and --keep-temp-dir
```
2019-10-28 12:22:17 ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 100G --CPU 10 --output '/tmp/atram_h76v8cy9/FHL.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_3ea5tkbi/trinity' --full_cleanup --no_bowtie' returned non-zero exit status 255.
```

**Comment by juliema** [comment]
you are running this on a high performance cluster? They usually require tat you load in the modules you are going to use e.g. atram, blast . trinity.

maybe yours is not set up that way.



On Mon, Oct 28, 2019 at 12:24 PM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

ok just totally checking here did you module load trinity and spades?

no, I don't know what that means.

Still get the assembler error when removing --tmp and --keep-temp-dir

2019-10-28 12:22:17 ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 100G --CPU 10 --output '/tmp/atram_h76v8cy9/FHL.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_3ea5tkbi/trinity' --full_cleanup --no_bowtie' returned non-zero exit status 255.

**Comment by juliema** [comment]
could this be an issue where your cluster is not letting the program write files?  it seems like these files are not being written to go into the programs.  are you getting any output from atram?

On Mon, Oct 28, 2019 at 12:26 PM Julie Allen <jallen23@unr.edu<mailto:jallen23@unr.edu>> wrote:
you are running this on a high performance cluster? They usually require tat you load in the modules you are going to use e.g. atram, blast . trinity.

maybe yours is not set up that way.



On Mon, Oct 28, 2019 at 12:24 PM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

ok just totally checking here did you module load trinity and spades?

no, I don't know what that means.

Still get the assembler error when removing --tmp and --keep-temp-dir

2019-10-28 12:22:17 ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 100G --CPU 10 --output '/tmp/atram_h76v8cy9/FHL.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_3ea5tkbi/trinity' --full_cleanup --no_bowtie' returned non-zero exit status 255.

**Comment by juliema** [comment]
actually it seems like a temporary file issue. I wonder if there are restrictions on the tmp/ files

On Mon, Oct 28, 2019 at 12:27 PM Julie Allen <jallen23@unr.edu<mailto:jallen23@unr.edu>> wrote:
could this be an issue where your cluster is not letting the program write files?  it seems like these files are not being written to go into the programs.  are you getting any output from atram?

On Mon, Oct 28, 2019 at 12:26 PM Julie Allen <jallen23@unr.edu<mailto:jallen23@unr.edu>> wrote:
you are running this on a high performance cluster? They usually require tat you load in the modules you are going to use e.g. atram, blast . trinity.

maybe yours is not set up that way.



On Mon, Oct 28, 2019 at 12:24 PM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

ok just totally checking here did you module load trinity and spades?

no, I don't know what that means.

Still get the assembler error when removing --tmp and --keep-temp-dir

2019-10-28 12:22:17 ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 100G --CPU 10 --output '/tmp/atram_h76v8cy9/FHL.aTRAM_SD.male_carcass_limited_SD_TCALIF_00592_1.fasta_01_3ea5tkbi/trinity' --full_cleanup --no_bowtie' returned non-zero exit status 255.

**Comment by etwatson** [comment]
> you are running this on a high performance cluster? They usually require tat you load in the modules you are going to use e.g. atram, blast . trinity.

No, I am trying this on my local machine. Now I see. I don't need a module load with assemblers on the HPC, since they aren't installed. I do module load python, etc..., but assemblers I have installed and in my path. 

I can try to troubleshoot on the HPC

**Comment by etwatson** [comment]
OK, I tried again on the HPC cluster with --temp-dir and --keep-temp-dir removed. It seems to 
have worked through several iterations before throwing an error:
"Exception: file is encrypted or is not a database"

```
---------- 2019-10-28 15:08:38 ---------
2019-10-28 15:08:41 INFO: ################################################################################
2019-10-28 15:08:41 INFO: aTRAM version: v2.3.0
2019-10-28 15:08:41 INFO: Python version: 3.7.4 (default, Jul 16 2019, 16:40:45) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
2019-10-28 15:08:41 INFO: /home/rcf-proj2/ew2/ericwats/repo/aTRAM/atram.py --db /home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM --probe /home/rcf-proj2/ew2/ericwats/probes/SD.testes_limited.fa --output-prefix testis --assembler trinity --iterations 25 --cpus 10 --ins-length 400
2019-10-28 15:08:41 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "SD.testes_limited.fa", iteration 1
2019-10-28 15:08:41 INFO: Blasting query against shards: iteration 1
2019-10-28 15:11:43 INFO: All 74 blast results completed
2019-10-28 15:11:43 INFO: 834 blast hits in iteration 1
2019-10-28 15:11:43 INFO: Writing assembler input files: iteration 1
2019-10-28 15:12:30 INFO: Assembling shards with trinity: iteration 1
2019-10-28 15:12:44 INFO: Saving assembled contigs: iteration 1
2019-10-28 15:12:47 INFO: Creating new query files: iteration 1
2019-10-28 15:12:47 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_01_long_reads.fasta", iteration 2
2019-10-28 15:12:47 INFO: Blasting query against shards: iteration 2
2019-10-28 15:13:10 INFO: All 74 blast results completed
2019-10-28 15:13:10 INFO: 599 blast hits in iteration 2
2019-10-28 15:13:10 INFO: Writing assembler input files: iteration 2
2019-10-28 15:13:28 INFO: Assembling shards with trinity: iteration 2
2019-10-28 15:13:38 INFO: Saving assembled contigs: iteration 2
2019-10-28 15:13:38 INFO: Creating new query files: iteration 2
2019-10-28 15:13:38 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_02_long_reads.fasta", iteration 3
2019-10-28 15:13:38 INFO: Blasting query against shards: iteration 3
2019-10-28 15:13:50 INFO: All 74 blast results completed
2019-10-28 15:13:50 INFO: 928 blast hits in iteration 3
2019-10-28 15:13:50 INFO: Writing assembler input files: iteration 3
2019-10-28 15:14:03 INFO: Assembling shards with trinity: iteration 3
2019-10-28 15:14:14 INFO: Saving assembled contigs: iteration 3
2019-10-28 15:14:15 INFO: Creating new query files: iteration 3
2019-10-28 15:14:15 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_03_long_reads.fasta", iteration 4
2019-10-28 15:14:15 INFO: Blasting query against shards: iteration 4
2019-10-28 15:14:27 INFO: All 74 blast results completed
2019-10-28 15:14:27 INFO: 1328 blast hits in iteration 4
2019-10-28 15:14:27 INFO: Writing assembler input files: iteration 4
2019-10-28 15:15:05 INFO: Assembling shards with trinity: iteration 4
2019-10-28 15:15:20 INFO: Saving assembled contigs: iteration 4
2019-10-28 15:15:21 INFO: Creating new query files: iteration 4
2019-10-28 15:15:21 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_04_long_reads.fasta", iteration 5
2019-10-28 15:15:21 INFO: Blasting query against shards: iteration 5
2019-10-28 15:15:32 INFO: All 74 blast results completed
2019-10-28 15:15:32 INFO: 1718 blast hits in iteration 5
2019-10-28 15:15:32 INFO: Writing assembler input files: iteration 5
2019-10-28 15:16:28 INFO: Assembling shards with trinity: iteration 5
2019-10-28 15:16:47 INFO: Saving assembled contigs: iteration 5
2019-10-28 15:16:48 INFO: Creating new query files: iteration 5
2019-10-28 15:16:48 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_05_long_reads.fasta", iteration 6
2019-10-28 15:16:48 INFO: Blasting query against shards: iteration 6
2019-10-28 15:17:00 INFO: All 74 blast results completed
2019-10-28 15:17:00 INFO: 2080 blast hits in iteration 6
2019-10-28 15:17:00 INFO: Writing assembler input files: iteration 6
2019-10-28 15:17:52 INFO: Assembling shards with trinity: iteration 6
2019-10-28 15:18:09 INFO: Saving assembled contigs: iteration 6
2019-10-28 15:18:10 INFO: Creating new query files: iteration 6
2019-10-28 15:18:10 INFO: aTRAM blast DB = "/home/rcf-proj2/ew2/ericwats/aTRAM/AB.aTRAM", query = "AB.aTRAM_SD.testes_limited.fa_06_long_reads.fasta", iteration 7
2019-10-28 15:18:10 INFO: Blasting query against shards: iteration 7
2019-10-28 15:18:12 ERROR: Exception: file is encrypted or is not a database

**Comment by rafelafrance** [comment]
On your current machine. Would you fire up a python3 console and
```python
import sqlite3
sqlite3.sqlite_version
```
Mine is saying 3.27.2
is your version even close to that?

**Comment by etwatson** [comment]
On the local machine, my version is 3.28.0
On the HPC cluster, the version is 3.7.17

**Comment by juliema** [comment]
ok so that is important I think we are troubleshooting two things here. Lets try to separate out the cluster issues with your local machine.  Rafe do you think they need to update the cluster?

On Tue, Oct 29, 2019 at 9:32 AM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

On the local machine, my version is 3.28.0
On the HPC cluster, the version is 3.7.17

**Comment by etwatson** [comment]
OK, that's positive. I just read that there are over 1 trillion sqlite3 databases in active use, so if versioning is causing an issue, it is a potentially huge one.

I'll stick to the HPC, since that affects our whole campus.

**Comment by rafelafrance** [comment]
I know that older version of SQLite3 have data integrity issues under heavy use, which is why I asked. I need to research which version are the issue etc. However, the 3.7.17 version is from May 2013, so yes, I'd recommend upgrading it.

At this point, we should start debugging things from the other end too. We need gather the bash scripts and some test data that fails. Of course we want to make the test data as small as possible with it still failing consistently.

**Comment by etwatson** [comment]
Apologies on the delay on my end. I submitted a ticket to my university asking them to update sqlite3 if possible 7 days ago and I haven't heard a thing. I've sent five follow up emails and still haven't even been contacted.

**Comment by juliema** [comment]
ok great thanks for keeping us posted.

On Mon, Nov 4, 2019 at 10:24 AM Eric Watson <notifications@github.com<mailto:notifications@github.com>> wrote:

Apologies on the delay on my end. I submitted a ticket to my university asking them to update sqlite3 if possible 7 days ago and I haven't heard a thing. I've sent five follow up emails and still haven't even been contacted.

**Comment by rafelafrance** [comment]
Ditto on keeping us posted. Thanks.

I don't know much about HPCs so, Question for Julie: Is sqlite3 a module you can load? And if you do would it be the one used by python? Just seeing if there's a way to bypass the IT department bureaucracy. I do believe is at least part of the problem I'm not 100% sure it's the only issue we're chasing here.

Eric: Question for you. Would it be possible to send us the scripts (along with some data) that are failing on your local machine?

**Comment by juliema** [comment]
agreed. and that is a good idea. Let me look into loading sqlite as a module on HPCs

On Mon, Nov 4, 2019 at 5:48 PM rafe <notifications@github.com<mailto:notifications@github.com>> wrote:

Ditto on keeping us posted. Thanks.

I don't know much about HPCs so, Question for Julie: Is sqlite3 a module you can load? And if you do would it be the one used by python? Just seeing if there's a way to bypass the IT department bureaucracy. I do believe is at least part of the problem I'm not 100% sure it's the only issue we're chasing here.

Eric: Question for you. Would it be possible to send us the scripts (along with some data) that are failing on your local machine?

**Comment by etwatson** [comment]
I called yesterday and they informed me that I need to install my own local sqlite3 on the HPC, so I am attempting that feat this morning. I'll work on sending some data and scripts to you today. Should I send just the raw data, and if so, what should I down sample the reads to since they're large files?

**Comment by rafelafrance** [comment]
First of all I want you thank you for your patience in dealing with the issue.

When you get the chance, I have two patches I want you to try on your local machine. You're getting fails there so let's use that one for this.

1. The first one is a simple `git pull` and run your test case. If that works then no real need to go on to step 2.

2. I did some major surgery on how I pass data between the main and subprocess threads.  I'd prefer not to use this code but if item 1 doesn't work then we'll have to try it. I put this code into a separate branch so to run it please check it out ```git checkout db_locked``` and then run your script.

**Comment by etwatson** [comment]
I tried both and neither worked. 

First patch:
- both assemblers I tried give this error here and there:
```
2019-11-08 11:17:23 ERROR: Exception: file is not a database
```

- spades gives:
```
2019-11-08 11:16:33 ERROR: The assembler failed with error: Command 'spades.py  --only-assembler --threads 10 --memory 100 --cov-cutoff off -o /tmp/atram_pda61blb/AB.aTRAM_SD.ovary_limited_SD_TCALIF_06058_1.fasta_01___q5xdly/spades' returned non-zero exit status 1
```
- trinity gives:
```
2019-11-08 11:16:56 ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 100G --CPU 10 --output '/tmp/atram_eegwu8hl/AB.aTRAM_SD.ovary_limited_SD_TCALIF_06058_1.fasta_01_mqjoljrs/trinity' --full_cleanup --no_bowtie' returned non-zero exit status 255.
```
- velvet doesn't give an error or do anything:
```
2019-11-08 11:21:01 INFO: No new assemblies in iteration 1
```

Second patch:
- trinity gives above error
- spades gives above error

**Comment by etwatson** [comment]
I ran ```git checkout master``` and tried running on a second dataset from a different species. This one may have issues with few blast hits with some genes. 

I still get  ```ERROR: Exception: file is not a database``` and ```ERROR: Exception: database disk image is malformed``` here and there.
I also got 
```
ERROR: The assembler failed with error: Command 'Trinity --seqType fa --max_memory 59G --CPU 20 --output '/tmp/atram_81mxs98c/SRI-SBMNH-1246_genes_CYTB_SRI_13.fasta_01_6wbf6b4a/trinity' --full_cleanup --no_bowtie --left '/tmp/atram_81mxs98c/SRI-SBMNH-1246_genes_CYTB_SRI_13.fasta_01_6wbf6b4a/paired_1.fasta' --right '/tmp/atram_81mxs98c/SRI-SBMNH-1246_genes_CYTB_SRI_13.fasta_01_6wbf6b4a/paired_2.fasta'' returned non-zero exit status 2.
```
once at the end, but otherwise got 8 tiny assemblies out of 13 genes.

**Comment by rafelafrance** [comment]
Let's look at the main database. If you could open that
```
sqlite3 path/to/<blast-db-name>.sqlite.db
sqlite> pragma quick_check;

or do the full but slower check

sqlite> pragma integrity_check;
```
